langwatch/src/components/checks/DynamicZodForm.tsx

```diff
@@ -35,6 +35,111 @@ import type { CheckConfigFormData } from "./CheckConfigForm";
 import { useOrganizationTeamProject } from "../../hooks/useOrganizationTeamProject";
 import { DEFAULT_EMBEDDINGS_MODEL, DEFAULT_MODEL } from "../../utils/constants";
 
+// Separate component for array fields to handle useFieldArray hook
+const ArrayField = <T extends EvaluatorTypes>({
+  fieldSchema,
+  fieldName,
+  evaluator,
+  variant = "default",
+  renderField,
+}: {
+  fieldSchema: ZodType;
+  fieldName: string;
+  evaluator: EvaluatorDefinition<T> | undefined;
+  variant?: "default" | "studio";
+  renderField: <T extends EvaluatorTypes>(
+    fieldSchema: ZodType,
+    fieldName: string,
+    evaluator: EvaluatorDefinition<T> | undefined
+  ) => React.JSX.Element | null;
+}) => {
+  const { control } = useFormContext();
+  const fullPath = fieldName;
+
+  const { fields, append, remove } = useFieldArray({
+    control,
+    name: fullPath,
+  });
+
+  // Cast to ZodArray to access element property
+  const arraySchema = fieldSchema as z.ZodArray<any>;
+
+  const defaultValues = useMemo(() => {
+    return arraySchema.element instanceof z.ZodObject
+      ? Object.fromEntries(
+          Object.entries(arraySchema.element.shape).flatMap(([key, value]) => {
+            if (value instanceof z.ZodUnion && value.options.length > 0) {
+              const defaultValue = value.options[0].value;
+              return [[key, defaultValue]];
+            }
+
+            return [];
+          })
+        )
+      : {};
```

Comment:
This seems surprisingly messy. I saw [this discussion](https://github.com/colinhacks/zod/discussions/1953).. but I don't know. It also is surprising to me that the assumption is that the defaults are hidden in the schema, instead of just having a prop for the defaults as would be a bit more standard?

i would think that since it's within the form context, the schema would be at the form level anyway, and that's where those defaults would come from? A lot of this code seems surprising to me, but if it works, it works.